### DISCUSSION TITLE
macOS Qt App Windowed mode displays terminal icon in dock.

### DISCUSSION THREAD

**Discussion by zoell** [pyinstaller/pyinstaller#6521] [Help] [answered]
I am trying to pack a PySide6 app which only displays a tray icon with a menu, no main window. I added the windowed mode parameter but when I start the binary (not the xxxx.app) as a user service for example using a simple plist file it displays a terminal icon in the dock. Wondering if it would be any way to not display any icon there?

**Comment by rokm** [comment, marked as answer]
The windowed mode primarily enables .app bundle generation, so it should make no difference.

Sounds like you want to make your application a background one, or an agent application. Which, from what I understand, is usually done via `Info.plist` - see [here](https://wiki.lazarus.freepascal.org/Hiding_a_macOS_app_from_the_Dock).

You can also transform your program to background/agent application at runtime by binding `ApplicationServices` library via `ctypes`; see the example below:

<details><summary>Example: run-time application transformation</summary>
<p>

```python
import os
import sys
import signal

from PySide6 import QtCore, QtWidgets, QtGui

# ctypes binding for ApplicationServices
import ctypes

class ProcessSerialNumber(ctypes.Structure):
    _fields_ = [
        ('highLongOfPSN', ctypes.c_uint32),
        ('lowLongOfPSN', ctypes.c_uint32),
    ]

kCurrentProcess = 2

kProcessTransformToForegroundApplication = 1
kProcessTransformToBackgroundApplication = 2
kProcessTransformToUIElementApplication = 4

app_services = ctypes.CDLL('/System/Library/Frameworks/ApplicationServices.framework/ApplicationServices')
app_services.TransformProcessType.argtypes = [ctypes.POINTER(ProcessSerialNumber), ctypes.c_uint32]
app_services.SetFrontProcess.argtypes = [ctypes.POINTER(ProcessSerialNumber)]

# Main
signal.signal(signal.SIGINT, signal.SIG_DFL)

app = QtWidgets.QApplication(sys.argv)

if not QtWidgets.QSystemTrayIcon.isSystemTrayAvailable():
    QtWidgets.QMessageBox.Warning(None, "Error", "System tray unavailable!")
    sys.exit(1)

quitAction = QtGui.QAction("Quit")
quitAction.triggered.connect(app.quit)

def hello():
    QtWidgets.QMessageBox.information(None, "Hello", "Hello world!")

helloAction = QtGui.QAction("Hello")
helloAction.triggered.connect(hello)

def transform_app(process_type):
    print(f"Transform: {process_type}")
    try:
        psn = ProcessSerialNumber(0, kCurrentProcess)
        app_services.TransformProcessType(psn, process_type)
    except Exception as e:
        QtWidgets.QMessageBox.warning(None, "Error", f"Failed to transform app: {e}")

transformToFrontAction = QtGui.QAction("Transform to ForegroundApplication")
transformToFrontAction.triggered.connect(lambda: transform_app(kProcessTransformToForegroundApplication))

transformToBackAction = QtGui.QAction("Transform to BackgroundApplication")
transformToBackAction.triggered.connect(lambda: transform_app(kProcessTransformToBackgroundApplication))

transformToUiElementAction = QtGui.QAction("Transform to UiElement")
transformToUiElementAction.triggered.connect(lambda: transform_app(kProcessTransformToUIElementApplication))

trayIconMenu = QtWidgets.QMenu()
trayIconMenu.addAction(helloAction)
trayIconMenu.addSeparator()
trayIconMenu.addAction(transformToFrontAction)
trayIconMenu.addAction(transformToBackAction)
trayIconMenu.addAction(transformToUiElementAction)
trayIconMenu.addSeparator()
trayIconMenu.addAction(quitAction)

pixmap = QtGui.QPixmap(64, 64)
pixmap.fill(QtGui.QColor.fromRgb(255, 255, 0))
icon = QtGui.QIcon(pixmap)

trayIcon = QtWidgets.QSystemTrayIcon()
trayIcon.setIcon(icon)
trayIcon.setContextMenu(trayIconMenu)
trayIcon.show()

# Set initial app type
transform_app(kProcessTransformToUIElementApplication)

app.exec()
```

</p></details>

The downside of the run-time transformation is that the dock icon is still visible from the moment the program is launched until the transformation is performed...

**Reply by zoell** [reply]
Cool thanks for the hint regarding the plist file. I did some research quickly and found if I add the <key>LSUIElement</key><true/> it works. But wondering if it is possible to add this parameter when running the pyinstaller command? I know I can add it to a spec file but for the spec file I need to run the pyinstaller command first. Is it not possible to add it in the pyinstaller command directly?

**Reply by rokm** [reply]
> Is it not possible to add it in the pyinstaller command directly?

No, not all options are mapped to the command-line interface. In this case, using the spec file is the way to go.

**Reply by zoell** [reply]
Okay thanks!

**Reply by zoell** [reply]
@rokm actually an interesting thing happened. I noticed that when I have any random macOS window in focus after 3-4 seconds the focus switches to my python app, is this something to do with the plist file? (this app actually has no main window, just the tray icon as mentioned) I am running an app with a plist file as open /Applications/myapp/myappname.app not sure if this could be an issue.

**Reply by rokm** [reply]
Doesn't seem to be happening here with my test app:
<details><summary>Test app</summary>
<p>

```python
import os
import sys
import signal

from PySide6 import QtCore, QtWidgets, QtGui

signal.signal(signal.SIGINT, signal.SIG_DFL)

app = QtWidgets.QApplication(sys.argv)
app.setQuitOnLastWindowClosed(False)

if not QtWidgets.QSystemTrayIcon.isSystemTrayAvailable():
    QtWidgets.QMessageBox.Warning(None, "Error", "System tray unavailable!")
    sys.exit(1)

quitAction = QtGui.QAction("Quit")
quitAction.triggered.connect(app.quit)

def hello():
    QtWidgets.QMessageBox.information(None, "Hello", "Hello world!")

helloAction = QtGui.QAction("Hello")
helloAction.triggered.connect(hello)

trayIconMenu = QtWidgets.QMenu()
trayIconMenu.addAction(helloAction)
trayIconMenu.addSeparator()
trayIconMenu.addAction(quitAction)

pixmap = QtGui.QPixmap(64, 64)
pixmap.fill(QtGui.QColor.fromRgb(255, 255, 0))
icon = QtGui.QIcon(pixmap)

trayIcon = QtWidgets.QSystemTrayIcon()
trayIcon.setIcon(icon)
trayIcon.setContextMenu(trayIconMenu)
trayIcon.show()

app.exec()
```

</p>
</details>

Built as onedir .app bundle, with `LSUIElement` inserted into bundle's `Info.plist` via spec file, and then started with `open program.app`.

**Reply by zoell** [reply]
When I start myappname.app from a plist service I have to use the open -W -g myappname.app in order to make it work correctly, looks like all good now. Thanks!
